lib/wibox/layout/ratio.lua

```
@@ -0,0 +1,362 @@
+--   end result is within 0 and 1
+function ratio:inc_widget_ratio(widget, increment)
+    if not widget or not increment then return end
+
+    local index = self:index(widget)
+
+    self:inc_ratio(index, increment)
+end
+
+--- Set the ratio of the widget at position `index`
+-- @tparam number index The index of the widget to change
+-- @tparam number percent An floating point value between 0 and 1
+function ratio:set_ratio(index, percent)
+    if not percent or #self.widgets ==  1 or not index or not self.widgets[index]
+        or percent < 0 or percent > 1 then
+        return
+    end
+
+    local old = self._ratios[index]
+
+    -- Remove what has to be cleared from all widget
+    local delta = ( (percent-old) / (#self.widgets-1) )
+
+    for k, v in pairs(self.widgets) do
+        self._ratios[k] = self._ratios[k] - delta
+    end
+
+    -- Set the new ratio
+    self._ratios[index] = percent
+
+    -- As some widgets may now have a slightly negative ratio, normalize again
+    normalize(self)
+
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Set the ratio of "widget"
+-- @param widget The widget to ajust
+-- @tparam number percent An floating point value between 0 and 1
+function ratio:set_widget_ratio(widget, percent)
+    local index = self:index(widget)
+
+    self:set_ratio(index, percent)
+end
+
+--- Update all widgets to match a set of a ratio.
+-- The sum of before, itself and after must be 1 or nothing will be done
+-- @tparam number index The index of the widget to change
+-- @tparam number before The sum of the ratio before the widget
+-- @tparam number itself The ratio for "widget"
+-- @tparam number after The sum of the ratio after the widget
+function ratio:ajust_ratio(index, before, itself, after)
+    if not self.widgets[index] or not before or not itself or not after then
+        return
+    end
+
+    local sum = before + itself + after
+
+    if sum > 1.01 or sum < -0.01 then return end
```

What's the purpose of this? It silently does nothing when.... something?